src/pico/i_video.c

```diff
@@ -949,108 +797,102 @@ void __noinline new_frame_init_overlays_palette_and_wipe() {
-#if !DEMO1_ONLY
-        video_scroll = next_video_scroll; // todo does this waste too much space
-#endif
-        sem_release(&display_frame_freed);
-    } else {
+    sem_acquire_blocking(&render_frame_ready);
+    display_video_type = next_video_type;
+    display_frame_index = next_frame_index;
+    display_overlay_index = next_overlay_index;
 #if !DEMO1_ONLY
-        video_scroll = NULL;
+    video_scroll = next_video_scroll; // todo does this waste too much space
 #endif
-    }
+    sem_release(&display_frame_freed);
     if (display_video_type != VIDEO_TYPE_SAVING) {
         // this stuff is large (so in flash) and not needed in save move
         new_frame_init_overlays_palette_and_wipe();
     }
 }
 
+// this is the code that sorts the vpatches and passes them to draw_vpatch, which directly draws the relevant pixels of the overlay into the buffer
+void __scratch_x("scanlines") handle_overlays(uint16_t *buffer, int scanline) {
+    assert(scanline < count_of(vpatchlists->vpatch_starters));
+    int prev = 0;
+    for (int vp = vpatchlists->vpatch_starters[scanline]; vp;) {
+        int next = vpatchlists->vpatch_next[vp];
+        while (vpatchlists->vpatch_next[prev] && vpatchlists->vpatch_next[prev] < vp) {
+            prev = vpatchlists->vpatch_next[prev];
+        }
+        assert(prev != vp);
+        assert(vpatchlists->vpatch_next[prev] != vp);
+        vpatchlists->vpatch_next[vp] = vpatchlists->vpatch_next[prev];
+        vpatchlists->vpatch_next[prev] = vp;
+        prev = vp;
+        vp = next;
+    }
+    vpatchlist_t *overlays = vpatchlists->overlays[display_overlay_index];
+    prev = 0;
+    for (int vp = vpatchlists->vpatch_next[prev]; vp; vp = vpatchlists->vpatch_next[prev]) {
+        patch_t *patch = resolve_vpatch_handle(overlays[vp].entry.patch_handle);
+        int yoff = scanline - overlays[vp].entry.y;
+        if (yoff < vpatch_height(patch)) {
+            vpatchlists->vpatch_doff[vp] = draw_vpatch((uint16_t*)(buffer), patch, &overlays[vp],
+                                                        vpatchlists->vpatch_doff[vp]);
+            prev = vp;
+        } else {
+            vpatchlists->vpatch_next[prev] = vpatchlists->vpatch_next[vp];
+        }
+    }
+}
+
 void __scratch_x("scanlines") fill_scanlines() {
-#if SUPPORT_TEXT
-    struct scanvideo_scanline_buffer *buffer = scanvideo_begin_scanline_generation_linked(display_video_type == VIDEO_TYPE_TEXT ? 2 : 1, false);
-#else
-    struct scanvideo_scanline_buffer *buffer = scanvideo_begin_scanline_generation(false);
-#endif
+    frame++;
+    // this shouldn't be here, but it's required for some reason
+    if (display_video_type != VIDEO_TYPE_SAVING && frame > 1) {
+        // this stuff is large (so in flash) and not needed in save move
+        new_frame_init_overlays_palette_and_wipe();
+    }
+
+
 #if USE_INTERP
     need_save = interp_in_use;
     interp_updated = 0;
 #endif
+    uint16_t buffer[SCREENWIDTH];
+    I_handleFrameStart(frame);
```

Comment:
and here's where the screen handling code hooks into pd_render. This function is defined in screen.c